doc/rsp.md

```diff
@@ -280,95 +315,168 @@ mapped into the 16-bit lanes. Signed opcodes (`LPV`, `SPV`) map the value to bit
-to 64 bits.
+register `VS<vs_elem>`, while `mtc2` moves `VS<vs_elem>` to GPR `rt`, sign
+extending to 64 bits.
 
 `ctc2` moves the lower 16 bits of GPR `rt` into the control register
 specified by `vs`, while `cfc2` does the reverse, moving the control register
 specified by `vs` into GPR `rt`, sign extending to 64 bits. Note that both
-`ctc2` and `cfc2` ignore the `vs_elem` field. For these instructions, the
+`ctc2` and `cfc2` ignore the `elem` field. For these instructions, the
 control register is specified as follows:
 
 | `vs` | Register |
-|---|---|
+| --- | --- |
 | 0 | `VCO` |
 | 1 | `VCC` |
 | 2 | `VCE` |
 
 
 Single-lane instructions
 ========================
+
 | 31..26 | 25 | 24..21 | 20..16 | 15..11 | 10..6 | 5..0 |
 | --- | --- | --- | --- | --- | --- | --- |
-| `COP2`| 1 | `vd_elem` | `vt` | `vt_elem` | `vd` | `opcode` |
+| `COP2`| 1 | `vt_elem` | `vt` | `vd_elem` | `vd` | `opcode` |
 
 Single-lane instructions are an instruction group that perform operations on a
-single lange of a single input register (`VT<se>`), and store the result into a single
-lane of a single output register (`VD<de>`).
-
-`vt_elem` and `vd_elem` are used to compute `se` and `de` that is to specify which lane,
-respectively of the input and output register, is affected.
-
-`vd_elem` is 4 bits long (range 0..15); the highest bit is always ignored so
-the destination lane `de` is computed from the lowest 3 bits.
-
-`vt_elem` is 5 bits long (range 0..31). `vt_elem(4)` must be zero. When
-`vt_elem(3)` is 1, `vt_elem(2..0)` is actually used as source lane `se`, as expected. When
-`vt_elem(3)` is 0, a hardware bug is triggered and portions of the lower bits of
-`vt_elem` are replaced with portion of the bits of `vd_elem` while computing `se`. Specifically, all
-bits in `vt_elem` from the topmost set bit and higher are replaced with the
-same-position bits in `vt_elem`. Notice that this behaviour is actually consistent
-with what happens when `vt_elem(3)` is 1, which means that there is no need to
-think of it as a special-case. Pseudo-code:
-
-    de(2..0) = vd_elem(2..0)
-    msb = highest_set_bit(vt_elem)
-    se(2..0) = vd_elem(2..msb) || vt_elem(msb-1..0)
-
-TODO: complete analysis for `vt_elem(4)` == 1.
+single lange of a single input register (`VT<vt_elem>`), and store the result into
+a single lane of a single output register (`VD<vd_elem>`). Only the lowest 3
+bits of `vt_elem` and `vd_elem` are used to compute the source lane `se` and destination
+lane `de`, respectively.
+
+For all single-lane instructions, `vt_elem` is further used as a "broadcast modifier",
+(as in other SIMD architectures) that modifies `vt`, duplicating some lanes and hiding
+others. All 4 bits of `vt_elem` are used for this purpose. Note that the broadcast
+modifier `vt_elem` affects the value of `VT` whenever it is used within an instruction,
+but these modifications are not written back - that is, from the point of view of other
+instructions the value of `VT` remains the same.
```

Comment:
Good catch! I had not realized that the `VMOV` access pattern was the same of the broadcast modifiers.

This said, are you sure this applies to *all* single-lane instructions? When I look at my code and my tests, I would say it only applies to `VMOV`. Can you provide a small snippet that shows this for any other single-large opcode?

Reply:
Thank you so much for reviewing my additions, I really appreciate your insight.

If I understand you correctly, I don't think this contradicts your code or tests. Since lane `elem & 7` stays the same after applying broadcast modifier `elem`, I don't think it's possible to distinguish whether the modifier is applied or not. I suppose I thought it was simpler to describe it as applying to all the instructions.

Reply:
I'm still missing something, then. You're right that it's impossible to distinguish whether the broadcast modifiers is applied or not because `elem & 7` stays the same.

But then, this doesn't explain the hardware bug that I described here (in the part you're removed). That hardware bug mixes `vt_elem` and `vd_elem` together; it's true that the way they're mixed is somewhat related to the way broadcast modifier works, but it's still a behavior which is specific of `VMOV` and requires a special description, as far as I can tell.

Am I missing something?